### PR TITLE
add sql support in local mode & add sort support

### PR DESCRIPTION
**Pull request by houqp**
Build of this PR depends on https://github.com/apache/arrow/pull/7464.

### CODE REVIEW

File: rust/ballista/src/dataframe.rs [hunk 1]
```diff
@@ -91,12 +96,52 @@ impl Configs {
     }
 }
 
+#[derive(Debug)]
+pub struct ContextSchemaProvider {
+    pub temp_tables: HashMap<String, DataFrame>,
```
Comment (houqp) [inline rust/ballista/src/dataframe.rs:101]: This implementation is a little bit different from the one in datafusion, it allows registering arbitrary dataframe as a temporary table for access within sql query.

File: rust/ballista/src/dataframe.rs [hunk 2]
```diff
@@ -91,12 +96,52 @@ impl Configs {
...
+
+#[derive(Debug)]
 pub struct Context {
-    state: Arc<ContextState>,
+    pub state: Arc<ContextState>,
+}
+
+#[derive(Debug)]
+pub struct ContextState {
+    pub schema_provider: RwLock<ContextSchemaProvider>,
```
Comment (houqp) [inline rust/ballista/src/dataframe.rs:139]: this reference needs to be mutable due to the need to reconstruct context from any dataframe objection. it also needs to be Send due to async :(

File: rust/ballista/src/dataframe.rs [hunk 3]
```diff
@@ -197,47 +271,46 @@ fn parse_settings(settings: HashMap<&str, &str>) -> HashMap<String, String> {
...
 }
 
 impl DataFrame {
     /// Create a builder from an existing plan
-    pub fn from(ctx: Arc<ContextState>, plan: &LogicalPlan) -> Self {
-        Self {
-            ctx_state: ctx,
-            plan: plan.clone(),
-        }
+    pub fn from(ctx_state: Arc<ContextState>, plan: LogicalPlan) -> Self {
```
Comment (houqp) [inline rust/ballista/src/dataframe.rs:284]: Remove reference in plan to push clone responsibility to caller. There are cases where clone is not needed if caller intend to transfer ownership entirely.

**Comment by houqp** [comment]
rebased to pull in fix from https://github.com/apache/arrow/pull/7464.

**Comment by andygrove** [comment]
@houqp Could you fix the conflict?

**Comment by houqp** [comment]
@andygrove rebased and builds are passing now.

**Comment by andygrove** [comment]
Thanks @houqp

**Pull request merged**